File: lsb-py/tests/test_0.py
import unittest
import lsb_py
from pathlib import Path

class TestLsbPy(unittest.TestCase):
    def setUp(self):
        self.path = Path(__file__).parent
        self.input = self.path / ".." / ".." / "data" / "input.webp"
        self.container = self.path / ".." / ".." / "data" / "container.webp"
        self.embedded = self.path / ".." / ".." / "data" / "embedded.png"

    def test_embed(self):
        with self.input.open("rb") as f:
            input_data = f.read()
        with self.container.open("rb") as f:
            container_data = f.read()

        extension = self.input.suffix[1:]
        result = lsb_py.embed(input_data, extension, container_data)

        self.assertIsInstance(result, bytes)
        self.assertNotEqual(result, container_data)

    def test_extract(self):
        with self.embedded.open("rb") as f:
            embedded_data = f.read()

        result, format = lsb_py.extract(embedded_data)

        self.assertIsInstance(result, bytes)
        self.assertEqual(format, "webp")

    def test_embed_extract(self):
        with self.input.open("rb") as f:
            input_data = f.read()
        with self.container.open("rb") as f:
            container_data = f.read()

        extension = self.input.suffix[1:]
        embedded_data = lsb_py.embed(input_data, extension, container_data)
        result, format = lsb_py.extract(embedded_data)

        self.assertEqual(result, input_data)
        self.assertEqual(format, "webp")


if __name__ == "__main__":
    unittest.main()

File: lsb-core/tests/test_0.rs
use lsb_core::{hash::Hash, *};

const INPUT: &[u8] = include_bytes!("../../data/input.webp");
const CONTAINER: &[u8] = include_bytes!("../../data/container.webp");
const EMBEDDED: &[u8] = include_bytes!("../../data/embedded.png");

#[test]
fn test_embed() {
    let hash = Hash::Sha256;
    let seed = 42;
    let lsbs = 1;
    let format = image::ImageFormat::WebP;

    let result = embed(INPUT, "webp", CONTAINER, lsbs, hash, seed, format);

    assert!(result.is_ok(), "Failed to embed data: {:?}", result.err());
}

#[test]
fn test_extract() {
    let seed = 42;
    let lsbs = 1;

    let extracted_result = extract(EMBEDDED, lsbs, seed);

    assert!(
        extracted_result.is_ok(),
        "Failed to extract data: {:?}",
        extracted_result.err()
    );
}

#[test]
fn test_embed_extract() -> Result<(), Box<dyn std::error::Error>> {
    let hash = Hash::Sha256;
    let seed = 42;
    let lsbs = 1;
    let format = image::ImageFormat::WebP;

    let embedded_result = embed(INPUT, "webp", CONTAINER, lsbs, hash, seed, format);

    assert!(
        embedded_result.is_ok(),
        "Failed to embed data: {:?}",
        embedded_result.err()
    );

    let embedded_data = embedded_result?;

    let extracted_result = extract(&embedded_data, lsbs, seed);

    assert!(
        extracted_result.is_ok(),
        "Failed to extract data: {:?}",
        extracted_result.err()
    );

    let extracted_data = extracted_result?.0;

    assert_eq!(
        INPUT, &extracted_data,
        "Extracted data does not match input"
    );

    Ok(())
}
